### PR TITLE
Change packaging to support Ubuntu 17.10 and 18.04

### PR DESCRIPTION
**Pull request by rjmholt**
## PR Summary

Ubuntu 17.04 is EOL, so we need to update our packaging for Ubuntu 17.10. This updates the package script and `build.psm1` to use 17.10 rather than 17.04.

The other things we should change to fully migrate are:
- The dpkg URL and name
- The README's [Get PowerShell instructions](https://github.com/PowerShell/PowerShell/blob/master/README.md#get-powershell)
- The Ubuntu 17.04 [installation instructions](https://github.com/PowerShell/PowerShell/blob/master/docs/installation/linux.md#ubuntu-1704)
- The Docker Ubuntu 17.10 [template](https://github.com/PowerShell/PowerShell/blob/8aec1585195e6d4871538aaa2ad2b112cf5fbfee/docker/tests/Templates/ubuntu17.10/Dockerfile#L24)

Also, Fedora 25 is now EOL and we should move to 27. This should already be handled, but we need to test on in a Fedora 27 environment.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by rjmholt** [comment]
Installing PowerShell on Fedora 28 I notice two things so far:
- The installation manager GUI window you get when you double click on the RPM to install the PowerShell binary says the license is "proprietary" rather than "MIT".
- I run into #6132.

**Comment by iSazonov** [comment]
I think we should rebase to move to .Net Core 2.1 and then chack that #6132 was fixed.

**Comment by rjmholt** [comment]
28 test failures in my Fedora 27 VM:

**Comment by rjmholt** [comment]
Okay I've run successfully installed PowerShell from `master` and run `Start-PSPester -Tag "CI","Feature","RequireSudoOnLinux"` under `sudo` on the following platforms:

|     | Ubuntu 17.10 | Ubuntu 18.04 | Fedora 27 | Fedora 28 |
| :-: | :--------------: | :-------------: | :----------: | :----------: |
| Succeeded | 6226 | 6227 | 6226 | 6232 |
| Failed | 12 | 11 | 9 | 8 |
| Skipped | 1225 | 1225 | 1227 | 1227 |
| Pending | 419 | 419 | 419 | 419 |
| Inconclusive | 0 | 0 | 0 | 0 |
| **Total** | 7882
 | 7882 | 7881 | 7886 |
| Pester XML | [pester-tests-ubuntu17.10.xml.zip](https://github.com/PowerShell/PowerShell/files/1972827/pester-tests-ubuntu17.10.xml.zip) | [pester-tests-ubuntu18.04.xml.zip](https://github.com/PowerShell/PowerShell/files/1972828/pester-tests-ubuntu18.04.xml.zip) | [pester-tests-fed27.xml.zip](https://github.com/PowerShell/PowerShell/files/1972823/pester-tests-fed27.xml.zip) | [pester-tests-fed28.xml.zip](https://github.com/PowerShell/PowerShell/files/1972825/pester-tests-fed28.xml.zip) |

Some of the problems I found:
* On Fedora 28:
    - https://github.com/PowerShell/PowerShell/blob/090f8761e8a4c53045467c218100ea3404120a77/test/powershell/engine/Remoting/SSHRemotingAPI.Tests.ps1#L18 fails with the message `Cannot process argument transformation on parameter 'Skip'. Cannot convert value "System.Object[]" to type "System.Management.Automation.SwitchParameter". Boolean parameters accept only Boolean values and numbers, such as $True, $False, 1 or 0.` (line 27, char 102).
    - https://github.com/PowerShell/PowerShell/blob/f0b6ef2eb470c0866f21768547f34af4670467fd/test/powershell/Modules/Microsoft.PowerShell.Management/Get-Process.Tests.ps1#L4 fails with message `Expected regular expression {fedora} to match {root}, but it did not match.` (line 5) (`fedora` is the user on the VM).
    - https://github.com/PowerShell/PowerShell/blob/9ca059dcf8e21db106e82827dc492c611a7c55b6/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1#L182 fails with message `Expected an exception, with FullyQualifiedErrorId 'NewItemSymbolicLinkElevationRequired,Microsoft.PowerShell.Commands.NewItemCommand' to be thrown, but no exception was thrown.` (line 185).
    - https://github.com/PowerShell/PowerShell/blob/24cf0204421580b52b31213854737fdf9417963f/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1#L81 fails with message `Expected exactly {2001:4860:4860::8888}, but got {8.8.8.8}.` (line 84)
    - https://github.com/PowerShell/PowerShell/blob/24cf0204421580b52b31213854737fdf9417963f/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1#L93 fails with message `Expected exactly {2001:4860:4860::8888}, but got {8.8.8.8}.` (line 98)
    - https://github.com/PowerShell/PowerShell/blob/84344cbb32ec5e2479e00394951832794bf37244/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1#L185 fails with message `Expected {0}, but got {32}.` (line 195)
    - https://github.com/PowerShell/PowerShell/blob/cc3ffd8349547e968c00e0b175a6abe348087b41/test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Alias.Tests.ps1#L97 fails with message `Expected an exception, with FullyQualifiedErrorId 'FileOpenFailure,Microsoft.PowerShell.Commands.ExportAliasCommand' to be thrown, but no exception was thrown.` (line 108)
    - https://github.com/PowerShell/PowerShell/blob/cc3ffd8349547e968c00e0b175a6abe348087b41/test/powershell/Modules/Microsoft.PowerShell.Utility/Set-Date.Tests.ps1#L15 fails with message `Expected string to be the same, but they were different. ... Expected {System.ComponentModel.Win32Exception,Microsoft.PowerShell.Commands.SetDateCommand}, But Was: {No Exception!}` (line 77)'

* On Fedora 27, all the failures experienced in 28, plus a `BeforeAll` block:
    https://github.com/PowerShell/PowerShell/blob/1b3c8aca507de2984cb14f0780dde94d977c7b8d/test/powershell/engine/ParameterBinding/ParameterBinding.Tests.ps1#L374
    fails with `The file '/tmp/TestBindingCmdlet.dll' already exists.` (line 374, char 13).

* On Ubuntu 18.04, all the tests failing on Fedora 27 fail, plus:
    - a native command processor test seems to call `less`, which halts the test execution. This test:
https://github.com/PowerShell/PowerShell/blob/ceef2e37e93826300d6638c0294ce7b0a5af591b/test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1#L133
    - The above test also fails with message `Expected exactly {/tmp/f581d4d2-b000-43d7-8960-0c9de6081b30/TextFileTest.txt}, but got {}.` (line 210)
    - https://github.com/PowerShell/PowerShell/blob/24cf0204421580b52b31213854737fdf9417963f/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1#L31 fails with message `Expected {0}, but got {32}.` (line 46).
    - https://github.com/PowerShell/PowerShell/blob/cc3ffd8349547e968c00e0b175a6abe348087b41/test/powershell/Modules/Microsoft.PowerShell.Utility/Invoke-Item.Tests.ps1#L124 fails with message `Expected {/home/ubuntu/PowerShell/src/powershell-unix/bin/Linux/netcoreapp2.1/linux-x64/publish}, but got {}.` (line 148).

* On Ubuntu 17.10, all tests failing on Ubuntu18.04 fail, plus:
    - https://github.com/PowerShell/PowerShell/blob/24cf0204421580b52b31213854737fdf9417963f/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1#L150 fails with message `Expected {0}, but got {2}.` (line 154)

**Comment by rjmholt** [comment]
I built a .deb the other day for Ubuntu 17.10, but it didn't work. Trying again today.

**Comment by rjmholt** [comment]
Have verified a working Ubuntu 17.10 package built with this branch.

(Can't attach due to size)

**Comment by rjmholt** [comment]
Also confirm that the produced Ubuntu 18.04 package works!

File: tools/packaging/packaging.psm1 [hunk 1]
```diff
@@ -993,6 +995,7 @@ function Get-PackageDependencies
                 "ubuntu.14.04" { $Dependencies += @("libssl1.0.0", "libicu52") }
                 "ubuntu.16.04" { $Dependencies += @("libssl1.0.0", "libicu55") }
                 "ubuntu.17.10" { $Dependencies += @("libssl1.0.0", "libicu57") }
+                "ubuntu.18.04" { $Dependencies += @("libssl1.0.0", "libssl1.1", "libicu60") }
```
Comment (iSazonov) [inline tools/packaging/packaging.psm1]: Do we really need both libssl?

File: tools/packaging/packaging.psm1 [hunk 2]
```diff
@@ -992,7 +994,8 @@ function Get-PackageDependencies
             switch ($Distribution) {
                 "ubuntu.14.04" { $Dependencies += @("libssl1.0.0", "libicu52") }
                 "ubuntu.16.04" { $Dependencies += @("libssl1.0.0", "libicu55") }
-                "ubuntu.17.04" { $Dependencies += @("libssl1.0.0", "libicu57") }
+                "ubuntu.17.10" { $Dependencies += @("libssl1.0.0", "libicu57") }
+                "ubuntu.18.04" { $Dependencies += @("libssl1.0.0", "libssl1.1", "libicu60") }
```
Comment (TravisEz13) [inline tools/packaging/packaging.psm1]: I think @iSazonov  is correct.  We only need `libssl1.0.0`

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
@anmenaga FYI this will change the package name for ubuntu 17.10

**Comment by anmenaga** [comment]
@TravisEz13 good point, thx. I'll talk to @rjmholt to add update to dockerfiles.

**Comment by kenk667** [comment]
@rjmholt I may have an interim solution for installing on 18.04 without downgrading curl or libicu, I was able to successfully install on 18.04 by changing the dependencies for https://github.com/PowerShell/PowerShell/releases/download/v6.1.0-preview.2/powershell_6.1.0-preview.2-1.ubuntu.17.04_amd64.deb

```bash
dpkg-deb --raw-exatrct powershell_6.1.0-preview.2-1.ubuntu.17.04_amd64.deb 6.1.0_patched
```

Then alter two patterns, curl and libicu dependencies to current versions;

```bash
sudo awk '{gsub("curl3","curl4");gsub("libicu57","libicu60");print}' 6.1.0_pathched/DEBIAN/control
```

Repackage as a debian package;

```bash
dpkg-deb --build 6.1.0_pathched/ powershell_6.1.0-patched.2-1.ubuntu.18.04_amd64.deb
```

Now just install the newly patched .deb package, I like to use gdebi to install;

```bash
sudo gdebi powershell_6.1.0-patched.2-1.ubuntu.18.04_amd64.deb
```

I haven't extensively tested this but seems to be functional and stable so far

**Comment by TravisEz13** [comment]
@kenk667 The required libraries are determined by DotNet Core.  Changing the required libraries could leave you unsupported.

**Pull request merged**